telefunc/node/server/shield/codegen/generateShield.ts

```diff
@@ -0,0 +1,379 @@
+        `Set \`compilerOptions.strict\` to \`true\` in ${tsConfigFilePath} (needed for shield() generation)`
+      )
+    }
+
+    // This source file is used for evaluating the template literal types' values
+    project.createSourceFile(typeToShieldFilePath, getTypeToShieldSrc())
+  }
+
+  const project = projects[key]!
+  objectAssign(project, { tsConfigFilePath })
+
+  const shieldGenFilePath = path.join(
+    path.dirname(telefuncFilePath),
+    `__telefunc_shieldGen_${path.basename(telefuncFilePath)}`
+  )
+  const shieldGenSource = project.createSourceFile(shieldGenFilePath)
+  shieldGenSource.addImportDeclaration({
+    moduleSpecifier: getImportPath(shieldGenFilePath, typeToShieldFilePath),
+    namedImports: ['ShieldArrStr']
+  })
+
+  const files = project.getSourceFiles(telefuncFilePath)
+  assert(files.length <= 1)
+  const telefuncFileSource = files[0]
+  if (!telefuncFileSource) {
+    const sourceFiles = project.getSourceFiles().map(getSourceFilePath)
+    assert(false, { telefuncFileSource, sourceFiles })
+  }
+
+  return { project, telefuncFileSource, shieldGenSource }
+}
+
+function generate({
+  project,
+  telefuncFileSource,
+  shieldGenSource,
+  telefuncFilePath
+}: {
+  project: Project & { tsConfigFilePath: null | string }
+  telefuncFileSource: SourceFile
+  shieldGenSource: SourceFile
+  telefuncFilePath: string
+}): string {
+  const telefunctions = telefuncFileSource.getFunctions().filter((f) => f.isExported())
+  const telefunctionNames = telefunctions.flatMap((telefunction) => {
+    const name = telefunction.getName()
+    if (!name) return []
+    return [name]
+  })
+
+  shieldGenSource.addImportDeclaration({
+    moduleSpecifier: getTelefuncFileImportPath(telefuncFilePath),
+    namedImports: telefunctionNames
+  })
+
+  // assign the template literal type to a string
+  // then diagnostics are used to get the value of the template literal type
+  for (const telefunctionName of telefunctionNames) {
+    shieldGenSource.addTypeAlias({
+      name: getShieldName(telefunctionName),
+      type: `ShieldArrStr<Parameters<typeof ${telefunctionName}>>`
+    })
+  }
+
+  const shieldAlias = '__shieldGenerator_shield' // alias for shield
+  telefuncFileSource.addImportDeclaration({
+    moduleSpecifier: 'telefunc',
+    namedImports: [
+      {
+        name: 'shield',
+        alias: shieldAlias
+      }
+    ]
+  })
+  telefuncFileSource.addVariableStatement({
+    declarationKind: VariableDeclarationKind.Const,
+    declarations: [
+      {
+        name: tAlias,
+        initializer: `${shieldAlias}.type`
+      }
+    ]
+  })
+
+  // Add the dependent source files to the project
+  project.resolveSourceFileDependencies()
+
+  // We need `compilerOptions.strict` to avoid `TS2589: Type instantiation is excessively deep and possibly infinite.`
+  assert(project.compilerOptions.get().strict === true)
+
+  /*
+  console.log(project.getSourceFiles().map(getSourceFilePath))
```

Comment:
Can be removed?

Reply:
Yes. I left it has a little debug convenience. I know many people don't like this kind of debugging comments. I just removed it.